### PR TITLE
[#17965] Create token component

### LINKED ISSUE #17965: Create token component
For the wallet designs the token image assets are all stored locally. Rather than have a long hardcoded list of assets it would be better to create a component similar to icon component which handles the resolving of token images. These assets should be resolved by their symbol - eg "SNT"

In the case of no asset being available for a given token there is a fallback image - align with Desktop/design for this image.

Also there is a community token which are passed as a base64 string 

Consider creating a new folder for the assets if there is unused/old assets in the old tokens folder.

Acceptance criteria:
Token component to resolve by symbol, base64 string or fallback image.
Add some test coverage if possible.
Verify image assets are correct

### PR DESCRIPTION
**Pull request by ulisesmac**
fixes #17965

### Summary

This PR creates a new token component, a similar abstraction to our icon component.

Everytime we want to draw a token, we should use this new implementation.

### Features

This  new component is defined in `quo.components.token` and is also available in the quo library (same as `icon` component). I created it to be very flexible, since it'll replace existing token images in the app.

Receives:
 | prop | description |
 | --- | ---|
 | size | it can be a number or a keyword, e.g. `:size-24`, `:size-2000` or just `2000` |
 | token | a token symbol in any format: e.g. `:eth`, `:ETH`, `"ETH"`, `"eth"` or even `:my-token/eTh`, all of them resolve to the same.|
 | style | extra styles to apply |
 | image-source | if provided, then we don't need `token`, the component will use this value, useful to render b64 strings.  The issue:  #17873   will benefit from it.|
  
### Changes
  
Besides adding this component, it is being used in all places in the app where we rendered a token image.
  
So, inside the quo library the new component is being used in:
- `token-input`
- `network-amount`
- `token-tag`
- `summary-tag`
- `context-tag`
- `token-value`
- `token-network`
- `preview-list`
  
All of these components received an image-source for the token to render, in this PR their API has been changed since they don't need that image-source anymore; it also means **every component that used any of these components has also been updated in this PR**

Also the usages of these components outside quo library have been updated.

### Review notes

Some additional small changes were performed, just small refactors, I'll mention what they are in code comments below.

## QA Testing notes

Please test all token images in the app. All of them should render exactly as before. Keep in mind this issue is not solving #17873 , but it can be solved after this one.

#### Platforms

- Android
- iOS

status: ready

### CODE REVIEW

File: src/quo/components/community/token_gating.cljs [hunk 1]
```diff
@@ -10,40 +10,36 @@
...
-                     [rn/view {:style style/token-tag-spacing}
-                      [token-tag/view
-                       {:token-symbol  (:symbol token)
-                        :token-value   amount
-                        :size          :size-24
-                        :options       (cond
-                                         sufficient?  :hold
-                                         purchasable? :add)
-                        :token-img-src img-src}]]))
-                 tokens))])
```
Comment (ulisesmac) [inline src/quo/components/community/token_gating.cljs:28]: Small refactor: `doall` is not needed here since we are not using `deref` (`@`) inside a lazy seq

File: src/quo/components/community/token_gating.cljs [hunk 2]
```diff
@@ -10,40 +10,36 @@
...
+                       :options      (cond
+                                       sufficient?  :hold
+                                       purchasable? :add)}]]))
+                tokens)])
 
 (defn- internal-view
   [{:keys [tokens padding? theme]}]
   [:<>
    (if (> (count tokens) 1)
-     (doall
```
Comment (ulisesmac) [inline src/quo/components/community/token_gating.cljs:34]: Same doall reason as before:

> Small refactor: doall is not needed here since we are not using deref (@) inside a lazy seq

File: src/quo/components/tags/token_tag/view.cljs [hunk 3]
```diff
@@ -14,35 +15,38 @@
...
+    (fn [{:keys [options size blur? theme token-value token-symbol]
           :or   {size :size-24}}]
       [rn/view
        {:on-layout #(reset! container-width
                       (oget % :nativeEvent :layout :width))}
        [hole-view/hole-view
         {:holes (if options
                   [{:x            (- @container-width
-                                     (condp = size
+                                     (case size
```
Comment (ulisesmac) [inline src/quo/components/tags/token_tag/view.cljs:31]: just changed `condp =` to `case` since it's comparing against literals

File: src/quo/components/token.cljs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                     string?  (keyword (string/lower-case token))
+                     keyword? (keyword (string/lower-case (name token)))
+                     nil)]
+    (resources/get-token token-name)))
+
+(def ^:private get-token-image (memoize get-token-image*))
+(def ^:private b64-png-image-prefix "data:image/png;base64,")
+
+(defn view
+  "Render a token image.
```
Comment (ilmotta) [inline src/quo/components/token.cljs]: I'm not reviewing this PR @ulisesmac, but I felt like sharing that in the future we'll be replacing most docstrings in this style you wrote with schemas. All calls to view would then be verified automatically and it would be unlikely for it to become outdated.

Sometimes the docstring contains useful bits of info that can't be inferred from the schema, in which case adding a comment is sufficient (we do that a lot in status-go when defining structs).

The schema describing this `view` could be:

```clojure
(def ?schema
  [:=>
   [:map {:closed true}
    [:size {:optional true :default 32} pos-int?]
    [:token {:optional true :default :snt} :schema.common/token]
    [:style {:optional true} map?]
    ;; Ignores `token` and uses this as parameter to `rn/image`'s source.
    [:image-source {:optional true} :schema.common/image-source]
    ;; If true, adds `data:image/png;base64,` as prefix to the string passed as `image-source`.
    [:add-b64-prefix? {:optional true} boolean?]]
   :any])
```

And the schema for a token (registered in `schema.common` ns) could be:

```clojure
(def ?token
  (let [regex #"^[a-zA-Z]{2,30}$"]
    [:and
     [:or string? keyword?]
     [:fn {:error/message (str "should match regex " regex)}
      (fn [value]
        (re-find regex (name value)))]]))
```

cc @J-Son89, @clauxx, @yqrashawn

File: src/quo/components/token.cljs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+           :width  size-number
+           :height size-number)))
+
+(defn- get-token-image*
+  [token]
+  (let [token-name (condp #(%1 %2) token
+                     string?  (keyword (string/lower-case token))
+                     keyword? (keyword (string/lower-case (name token)))
+                     nil)]
+    (resources/get-token token-name)))
```
Comment (J-Son89) [inline src/quo/components/token.cljs]: oh I don't think we want to load from resources like this. It's better to use the macro so we can just pull directly from the resources/tokens folder and don't need to manually update the resources.cljs file with each token that we add. 
if anything we need to delete the entries in `resources.cljs` with respect to tokens 👍

File: src/quo/components/token.cljs [hunk 6]
```diff
@@ -0,0 +1,53 @@
+(ns quo.components.token
```
Comment (J-Son89) [inline src/quo/components/token.cljs]: maybe we can have this component in a specific namespace and put icon component there too? (we could move icon in a separate pr) 
e.g quo.components.utitlites or quo.components.asset-loaders or something like that?  the idea being to have a clearer separation that these files are not a direct DS component

File: src/quo/components/token.cljs (hunk 4 above)
Comment (J-Son89) [reply src/quo/components/token.cljs]: schema would be great to add here in place of doc string 👍

File: src/quo/components/token.cljs (hunk 6 above)
Comment (ulisesmac) [reply src/quo/components/token.cljs]: Atm they don't have a `view` ns, so that's the difference.

And yes, sounds good to move it, but the icon is being used in many places, maybe that's something not everyone will like.

It'd be good to have the opinion of other reviewers about it 👍

File: src/quo/components/token.cljs (hunk 4 above)
Comment (ulisesmac) [reply src/quo/components/token.cljs]: Will add it!

File: src/quo/components/token.cljs (hunk 5 above)
Comment (ulisesmac) [reply src/quo/components/token.cljs]: Interesting approach. I'll add it 👍

File: src/quo/components/token.cljs (hunk 4 above)
Comment (ulisesmac) [reply src/quo/components/token.cljs]: @ilmotta are we adding regex validations for all of our string/keyword props?

Sometimes we need it and that's ok, but I'd prefer not using them when it's not required.
A regex is very error prune (do we need tests for them?), IMO they are "kind of" easy to write, but hard to read in the future and also hard to maintain.

If we want to properly validate our strings/keywords, we'd need a more formal specification of what we can/can't receive.

For example, this token schema, what's the minimum length? why 2? what's the maximum? don't we have tokens returned from status-go that contain numbers or other symbols? who is able to answer these kind of questions for all the components we use?

On the other hand, if we create very flexible regexes, does that makes sense? I mean, what would be the benefit from it if the regex is not strict enough?

Comment (yqrashawn) [reply src/quo/components/token.cljs]: > all of these are valid and resolve to the same

The real issue seems to be why we need to support all these formats.

We can just follow the convention of what a token `symbol` usually looks like, all uppercase, numbers are ok, minimum length is 2, since symbol is actually optional in erc20 and just a string that can contain anything

File: src/quo/components/token.cljs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+   - token:           string or keyword, it can contain upper case letters or not.
+                      E.g. all of these are valid and resolve to the same:
+                      :token/snt | :snt | :SNT | \"SNT\" | \"snt\".
+   - image-source:    Ignores `token` and uses this as parameter to `rn/image`'s source.
+   - add-b64-prefix?: If true, adds `data:image/png;base64,` as prefix to the string
+                      passed as `image-source`.
+  "
+  [{:keys [token size style image-source add-b64-prefix?]
+    :or   {size  32
+           token :snt}}]
```
Comment (yqrashawn) [inline src/quo/components/token.cljs]: don't think it's a good idea to fallback to snt when token is not provided/defined
this may cause user confusion or even asset loss

File: src/quo/components/token.cljs (hunk 4 above)
Comment (ilmotta) [reply src/quo/components/token.cljs]: @ulisesmac, the code I suggested was just to help guide the final implementation if and only if you want to add the schemas in this PR. I should've been more clear. It's not something to be copy & pasted.

> @ilmotta are we adding regex validations for all of our string/keyword props?

It depends on how much we want to instrument and how specific we want them to be. There's no right or wrong. Our schemas can be quite dumb, focusing more on primitive types only, like sequences, strings, keywords, etc. But they can be much more useful than that because schemas are runtime constructs, we can check a lot more during instrumentation. More strict schemas are helpful for code generation, something we will eventually take advantage of.

Just a reminder, all the schema code we are writing is dev-only code, it should help us during development. If a schema is too complicated or doesn't help, we can throw away/change.

> On the other hand, if we create very flexible regexes, does that makes sense? I mean, what would be the benefit from it if the regex is not strict enough?

I just tried to convert the docstring you wrote into a valid schema and added a restriction in size to show the utility/power because most devs in the team haven't used malli yet. Don't worry too much about my code suggestion.

> If we want to properly validate our strings/keywords, we'd need a more formal specification of what we can/can't receive.

I agree, it's up to us to define that when working on the PRs. The schemas can get as "formal" or strict as we want them to be. We can start with just `[:or string? keyword?]` and evolve from there.

> A regex is very error prune (do we need tests for them?)

I think you're over generalizing regular expressions. Of course I agree they can be error prone, but the one I shared reads like English for many devs.

We don't need tests for schemas in general, unless they're much more complicated and used by production code. We are far away from that scenario. Instrumentation is dev-only code.

> We can just follow the convention of what a token symbol usually looks like, all uppercase, numbers are ok, minimum length is 2, since symbol is actually optional in erc20 and just a string that can contain anything

@yqrashawn seems reasonable 👍🏼

The fact that we're finally having a discussion about the types and formats of things is a good sign, we are thinking more about the data layer. That's kind of the whole point about using malli :)

File: src/quo/components/token.cljs (hunk 6 above)
Comment (ilmotta) [reply src/quo/components/token.cljs]: I would appreciate such structural clarity in the code @J-Son89, @ulisesmac. DS components should have a clear distinction from our own.

File: src/quo/components/token.cljs (hunk 4 above)
Comment (ulisesmac) [reply src/quo/components/token.cljs]: > > All of these are valid and resolve to the same
> 
> The real issue seems to be why we need to support all these formats.
> 
> We can just follow the convention of what a token `symbol` usually looks like, all uppercase, numbers are ok, minimum length is 2, since symbol is actually optional in erc20 and just a string that can contain anything

@yqrashawn

I'm supporting those formats because the token component is now being used directly or indirectly in many places in the app, replacing code that didn't expect the proper format. 

For example, our current (develop) code to resolve tokens expect a lower case keyword, and the new implementation should break those existing uses.

Regarding to the token's minimum length, I've noticed we have a token image saved as `R`, so our keyword to get it is `:r`.

Seems like I can't be too strict with this schema.

Comment (ulisesmac) [reply src/quo/components/token.cljs]: @ilmotta

Thanks for the detailed answer, I got the point. 👍

File: src/quo/components/token.cljs (hunk 7 above)
Comment (ulisesmac) [reply src/quo/components/token.cljs]: I agree, will remove it.

BTW, designers are pending to create a default component when no image is available

**Comment by ulisesmac** [comment]
@J-Son89  I've pushed the branch with the following changes:

- Moved the compone t to `quo.components.utilities.token.view`.
- All token files are being loaded on the app startup, using a macro with a similar apporach to our icons. There're two files:
`quo.components.utilities.token.loader` in clj and cljs to load the files.
-  The function `resources/get-token` has been removed along with its js/require map.

@ilmotta 
The schema has been added :+1: it's more flexible.

@yqrashawn 
`:snt` has been removed as default value for `:token`.

Thanks for your reviews :smiley:

**Review by mohsen-ghafouri — APPROVED** [review]
Well done  🚀, LGTM :)

**Review by smohamedjavid — APPROVED** [review]
Nice work! @ulisesmac 🚀

File: src/quo/core.cljs [hunk 8]
```diff
@@ -378,6 +379,9 @@
 (def text-combinations quo.components.text-combinations.view/view)
 (def channel-name quo.components.text-combinations.channel-name.view/view)
 
+;;;; Utilities
```
Comment (J-Son89) [inline src/quo/core.cljs]: can you leave a comment here - something like "outside of design system" 👍

**Review by J-Son89 — APPROVED** [review]
nice pr 🥇

**Review by briansztamfater — APPROVED** [review]
Nice work! 🚀  IMO we should request manual QA is this affects screens like select asset screen

File: src/quo/core.cljs (hunk 8 above)
Comment (ulisesmac) [reply src/quo/core.cljs]: Done!

**Comment by ulisesmac** [comment]
> Nice work! 🚀 IMO we should request manual QA is this affects screens like select asset screen

Thanks!
and yes! I wrote a note for QA in the description :)

**Review by yqrashawn — APPROVED** [review]
LGTM

**Review by mmilad75 — APPROVED** [review]
What a great solution 🚀

**Comment by churik** [comment]
@ulisesmac 
tests are failing with:
```

[2023-12-06T15:22:47.712Z] src/status_im2/contexts/wallet/common/temp.cljs:3:6: error: namespace quo.foundations.resources is required but never used

[2023-12-06T15:22:47.712Z] src/status_im2/contexts/wallet/common/temp.cljs:3:36: warning: Unused alias: quo.resources
```
can you please take a look?

**Comment by churik** [comment]
I checked this on all tokens that we have for our testing purposes, no difference - the only difference that I noticed is in sorting, take a look:
IMAGE 2023-12-06 18:08:52

Also, I checked components from the description + #17873
It fixes #17873 

IMAGE 2023-12-06 18:10:19
Several questions:
1) Is the issue with sorting related to your PR or it is just random everywhere?
2) What about the default fallback icon for a token (when no icon is set)?
3) Design review should be performed on this PR or not?

**Comment by ulisesmac** [comment]
@churik 
Thanks for the review, I was a little busy with other stuff, so I'll reply soon.

**Comment by ulisesmac** [comment]
Hey @churik 
 
> 1. Is the issue with sorting related to your PR or it is just random everywhere?

This is not related to my PR, but I checked the code and we are not sorting those tokens, so I'd say that right now, yes, it's random.

> 2. What about the default fallback icon for a token (when no icon is set)?

I asked to designers, and they said there's still no design for it, when it exists, we can add it :+1: 

> 3. Design review should be performed on this PR or not?

No, we don't need it :) It's just a code change  related.

This implementation supports b64 strings to be passed to the token component and also makes it easier to get the image of a token, so I can understand it's fixing
- #17873

But I'd say to properly ensure it's working is better to keep the issue open and address it separately.
@mohsen-ghafouri has it assigned, but I could also take it and check what's happening.

@mohsen-ghafouri wdyt?

**Comment by mohsen-ghafouri** [comment]
@ulisesmac I can see you already resolved it and we just need to check it for custom token. feel free and reassign that issue to yourself if you want. :)
